### PR TITLE
Add BoundingValue Damper and support for NodalDamper

### LINKED ISSUE #7856: Damper that bounds the values of a variable during solution
### Description of the enhancement or error report
Create a Damper that limits the value of a variable to be within user-prescribed minimum or maximum values.

### Rationale for the enhancement or information for reproducing the error
We commonly use the MaxIncrement Damper to prevent the temperature from going negative during solutions of the heat equation. This usually works, but picking a permissible temperature increment that prevents this issue but doesn't unduly interfere with the solution of the problem is an art. It would be better to simply specify the permissible minimum value of the temperature. Also, the MaxIncrement Damper is applied at quadrature points. It would be better to do this check at nodes.

### Identified impact
Because we would like to apply this to nodes rather than quadrature points, we would need to create a new NodalDamper base class, as well as a class for this specific damper.

### PR DESCRIPTION
**Pull request by bwspenc**
Closes #7856

### CODE REVIEW

**Comment by bwspenc** [comment]
A few comments/questions about this:
- The code in NonlinearSystem and MooseVariable for computing the increment needs to different things for nodal and element dampers. What I've done doesn't feel like a very elegant solution.
- The solution increment vector seems to actually be the negative of the solution increment. I could set the value of _increment to the negative of that value when I'm setting it, or use the negative of _increment when I'm using it if the sign matters in the damper. I did the latter, but it could be done either way.
- Both the BoundingValue and the MaxIncrement Damper could be implemented as either nodal or element dampers. I'm not sure whether it makes any sense to make versions of these that do it both ways, e.g. ElementBoundingValue, NodalBoundingValue. My opinion is that's probably getting a little carried away.
- I added a bit of code that prints to the console the value of the damping is if it's ever less than 1. That seems pretty helpful for the user to understand what's going on.
- I still need to add a test for this, which is a little tricky, but I have ideas. I wanted to get feedback on the code first.

File: framework/src/base/MooseVariable.C [hunk 1]
```diff
@@ -1605,6 +1605,18 @@ MooseVariable::computeIncrement(const NumericVector<Number> & increment_vec)
...
+void
+MooseVariable::computeIncrementAtNode(const NumericVector<Number> & increment_vec)
+{
+  if (!isNodal())
+    mooseError("computeIncrementAtNode can only be called for nodal variables");
+
+  _increment.resize(1);
+
+  // Compute the increment for the current DOF
+  _increment[0] =  increment_vec(_dof_indices[0]);
```
Comment (andrsd) [inline framework/src/base/MooseVariable.C]: one space after `=`

File: framework/src/base/NonlinearSystem.C [hunk 2]
```diff
@@ -692,6 +697,8 @@ NonlinearSystem::addDamper(const std::string & damper_name, const std::string &
     }
     else if (ed)
       _element_dampers.addObject(ed,tid);
+    else if (nd)
+      _nodal_dampers.addObject(nd,tid);
```
Comment (andrsd) [inline framework/src/base/NonlinearSystem.C]: This is a copy and paste, but add a space after `,`.

File: framework/src/base/ComputeNodalDampingThread.C [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  const Node * node = *node_it;
+  _fe_problem.reinitNode(node, _tid);
+
+  std::set<MooseVariable *> damped_vars;
+
+  const std::vector<MooseSharedPointer<NodalDamper> > & ndampers = _nl.getNodalDamperWarehouse().getActiveObjects(_tid);
+  for (const auto & damper : ndampers)
+    damped_vars.insert(damper->getVariable());
+
+  _nl.reinitIncrementAtNodeForDampers(_tid,damped_vars);
```
Comment (andrsd) [inline framework/src/base/ComputeNodalDampingThread.C]: Add a space after  `,`.

File: framework/src/base/ComputeElemDampingThread.C [hunk 4]
```diff
@@ -47,7 +47,14 @@ ComputeElemDampingThread::onElement(const Elem *elem)
...
   _fe_problem.reinitElem(elem, _tid);
-  _nl.reinitIncrementForDampers(_tid);
+
+  std::set<MooseVariable *> damped_vars;
+
+  const std::vector<MooseSharedPointer<ElementDamper> > & edampers = _nl.getElementDamperWarehouse().getActiveObjects(_tid);
+  for (const auto & damper : edampers)
+    damped_vars.insert(damper->getVariable());
+
+  _nl.reinitIncrementAtQpsForDampers(_tid,damped_vars);
```
Comment (andrsd) [inline framework/src/base/ComputeElemDampingThread.C]: A space after `,`

**Review by andrsd — CHANGES_REQUESTED** [review]

File: framework/src/base/MooseVariable.C (hunk 1 above)
Comment (bwspenc) [inline framework/src/base/MooseVariable.C]: Fixed

File: framework/src/base/NonlinearSystem.C (hunk 2 above)
Comment (bwspenc) [inline framework/src/base/NonlinearSystem.C]: Fixed

File: framework/src/base/ComputeNodalDampingThread.C (hunk 3 above)
Comment (bwspenc) [inline framework/src/base/ComputeNodalDampingThread.C]: Fixed

**Comment by bwspenc** [comment]
I addressed the comments of @andrsd and added a test.

File: test/tests/dampers/bounding_value/bounding_value_test.i [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+[]
+
+[Executioner]
+  type = Transient
+  solve_type = 'PJFNK'
+  end_time = 3.0
+#  [./TimeStepper]
+#    type = ConstantDT
+#    min_dt = 0.5
+#  [../]
```
Comment (andrsd) [inline test/tests/dampers/bounding_value/bounding_value_test.i]: Let's remove this commented out section

File: test/tests/dampers/bounding_value/bounding_value_test.i [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  xmax = 1
+  ymin = 0
+  ymax = 1
+  nx = 2
+  ny = 2
+  elem_type = QUAD9
+[]
+
+[Variables]
+  active = 'u'
```
Comment (andrsd) [inline test/tests/dampers/bounding_value/bounding_value_test.i]: I'd remove this

File: test/tests/dampers/bounding_value/bounding_value_test.i [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+  [./source]
+    type = BodyForce
+    variable = u
+    function = 't'
+  [../]
+[]
+
+[BCs]
+  active = 'left right'
```
Comment (andrsd) [inline test/tests/dampers/bounding_value/bounding_value_test.i]: Remove this as well

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by andrsd** [comment]
After cleaning up the input file, this looks good to me.

@friedmud Since you built dampers, I would not mind a second opinion on this.  What @bwspenc did is he mimicked what we do for elemental dampers, but he does it for nodal values.

**Comment by bwspenc** [comment]
@andrsd I made the changes you suggested to that test. While I was at it, I made the same changes to the test that I based it on, updated the comments at the top of the input file, and made a second version of the test that tests the min_value (the original one just tested max_value).

**Review by andrsd — APPROVED** [review]
I will let this sit here till like lunch. If nobody objects, I will merge then...

**Comment by dschwen** [comment]
Shouldn't this be called `BoundingValueNodalDamper` rather than `BoundingValue`? The latter is awfully general. We have an element damper in Marmot that does almost the same thing (but more specific) and could be replaced by a more general `BoundingValueElementDamper`.

**Comment by bwspenc** [comment]
@dschwen I was thinking the same thing (although I wouldn't put `Damper` at the end: how about `BoundingNodalValue`?).  Along those lines, it seems like the existing `MaxIncrement` Damper shoulde also be re-named (which will be a pain because it's so widely used), and that there probably should be a nodal version of that. That's an issue for another time, but I'm happy to re-name this.

**Comment by bwspenc** [comment]
@dschwen per our discussion, I renamed this to `BoundingValueNodalDamper` and added `BoundingValueElementDamper`, as well as versions of the tests for the new element version of this.

**Pull request merged**